### PR TITLE
Consistent service interfaces

### PR DESCRIPTION
**Pull request by marcelkornblum**
> This is a follow-on from #38  (and git branched from `simpler-services` used in that PR - please merge that one first).

Our services could be more consistent in the way they expose what is essentially an API; for example the function parameters they accept and the way they're named. We also need to do away with the use of untyped dictionaries as a transport type and essentially a catch-all.

Starting with the lowest level services I've reworked them to have more precise and concise function names (`create`, `delete` etc) and to consistently take a predictable set of parameters. By default the user service methods will want the first parameter to be a User object, for example. Since we're mainly going to be dealing with API requests that aren't authenticated to a specific User I've also included utility wrapper methods that accept an ID; these are all named consistently and use the same retrieval method.

I've surfaced the field names and values out of dictionaries and into individual function parameters so that the function signatures are clear about what is needed and what is optional; this also lets us add typing. Along with removing the extra layer of validation that seemed to be creeping in this means we have less surface area, fewer tests, clearer functionality, more consistent error reporting and more consistent style across codebases.

I've changed the return types of most of them as well; for the most part I am returning what the ORM returns; this is the `most Django` thing to do and there's no reason for us to be putting a layer of implicit types with all our various dictionary formats and different objects coming out what are essentially model methods.

Importantly I've also removed the use of broad-functioned internal helper methods that we don't seem to need; `get_or_create` doesn't seem the most appropriate method to use and it was skewing our approach. I've also removed most of the args_kwargs drilling since that was really only appropriate when lightly wrapping those powerful methods.

In some places (notably SCIM) all this has led to the services having nothing at all inside; if it's simply a wrapper around the core service (and why wouldn't it be) then there's no need at all for it and I've removed it entirely; our surface area is smaller as a result.

Note that a few things aren't working properly at the moment - specifically the SCIM interface/schema connection (to be addressed in a new PR)

> NB #40 depends on this PR

### CODE REVIEW

File: profiles/services/combined.py [hunk 1]
```diff
@@ -3,57 +3,54 @@
...
-def delete_profile(sso_email_id: str) -> Profile:
+def delete(profile: Profile):
     """
     Retrieve a user by their ID, only if the user is not soft-deleted.
     """
-    profile = Profile.objects.get(sso_email_id=sso_email_id)
     profile.is_active = False
-    profile.save()
-    return profile
+    return profile.save()
```
Comment (CamLamb) [inline profiles/services/combined.py:56]: `update_fields`

File: core/services.py [hunk 2]
```diff
@@ -11,13 +12,20 @@
...
+    Entrypoint for new user creation. Triggers the creation of User record,
+    then the relevant Profile record as well as a combined Profile.
+    """
+    user = user_services.create(sso_email_id=id)
+    if initiator == "SSO":
+        first_name = kwargs.get("first_name", None)
+        last_name = kwargs.get("last_name", None)
+        emails = kwargs.get("emails", None)
+        profile_services.create_from_sso(
+            user,
```
Comment (CamLamb) [inline core/services.py:26]:
```suggestion
            sso_email_id,
```

File: core/services.py [hunk 3]
```diff
@@ -11,13 +12,20 @@
...
-    return user_services.get_or_create_user(id, *args, **kwargs)
-
-
-def get_user_by_id(id: str) -> User:
-    return user_services.get_user_by_sso_id(id)
-
-
-# TODO:
-# - Add create_profile, get_profile etc. in this service
+def create_user(id: str, initiator: str = "SSO", **kwargs) -> User:
```
Comment (CamLamb) [inline core/services.py:15]:
```suggestion
def create_user(sso_email_id: str, initiator: str = "SSO", **kwargs) -> User:
```

File: core/services.py [hunk 4]
```diff
@@ -11,13 +12,20 @@
...
+    """
+    Entrypoint for new user creation. Triggers the creation of User record,
+    then the relevant Profile record as well as a combined Profile.
+    """
+    user = user_services.create(sso_email_id=id)
```
Comment (CamLamb) [inline core/services.py:20]:
```suggestion
    user = user_services.create(sso_email_id=sso_email_id)
```

File: core/services.py [hunk 5]
```diff
@@ -11,14 +12,20 @@
...
+    if initiator == "SSO":
+        first_name = kwargs.get("first_name", None)
+        last_name = kwargs.get("last_name", None)
+        emails = kwargs.get("emails", None)
```
[the remark this answers is not in the document]
Comment (CamLamb) [reply core/services.py:24]:
```python
def new_user(
    id: str,
    initiator: str,
    # TODO: DISCUSS: Should this be better typed as a TypedDict depending on which profile type
    # is being created?
    profile_data: dict,
) -> User:
    """
    Entrypoint for new user creation. Triggers the creation of User record,
    then the relevant Profile record as well as a combined Profile.
    """
    initiating_type = ProfileTypes(initiator)

    try:
        user = user_services.get_by_id(id)
    except User.DoesNotExist:
        user = user_services.create(sso_email_id=id)

    if initiating_type == ProfileTypes.STAFF_SSO:
        first_name: str = profile_data["first_name"]
        last_name: str = profile_data["last_name"]
        emails: list[dict] = profile_data["emails"]
        profile_services.create_from_sso(
            id,
            first_name,
            last_name,
            emails,
        )

    return user
    ```

[the remark this answers is not in the document]
Comment (marcelkornblum) [reply core/services.py:24]: perhaps the right answer is to stick to my heuristic and just make it work nicely for SSO which may be the only initiator we ever have...

File: core/services.py (hunk 2 above)
Comment (marcelkornblum) [reply core/services.py:26]: im confused by this; it's already done

File: profiles/services/staff_sso.py [hunk 6]
```diff
@@ -1,106 +1,157 @@
...
+# SSO Profile data methods
+###############################################################
+
+
+def get_by_user_id(id: str) -> StaffSSOProfile:
+    """
+    Retrieve a profile by its User ID, only if the user is not soft-deleted.
+    """
+    # @TODO update the below query (inc the model) to stop it requiring a JOIN
+    return StaffSSOProfile.objects.get(user_id=id, user__is_active=True)
```
[the remark this answers is not in the document]
Comment (marcelkornblum) [reply profiles/services/staff_sso.py:31]: I've done this for consistency but I think IU'd rather move them all to just being ID. Not really sure though (feels like we ought to abstract the name of the field away from the service API)

File: profiles/services/staff_sso.py [hunk 7]
```diff
@@ -1,106 +1,157 @@
...
-    staff_sso_profile_email = StaffSSOProfileEmail.objects.filter(email=email)[0]
-
-    staff_sso_profile_email.profile = profile
-    staff_sso_profile_email.email = email
-    staff_sso_profile_email.type = type
-    staff_sso_profile_email.preferred = preferred
-
-    staff_sso_profile_email.save()
-    return staff_sso_profile_email
+    staff_sso_email, _ = StaffSSOProfileEmail.objects.get_or_create(
```
[the remark this answers is not in the document]
Comment (marcelkornblum) [reply profiles/services/staff_sso.py:120]: I actually think this is a good time to use get_or_create - here's we're trying to make the association and create the email if it doesnt exist; I dont think the "already exists" at this point is a failure state...

File: profiles/services/staff_sso.py [hunk 8]
```diff
@@ -1,106 +1,157 @@
...
+    email_object, _ = Email.objects.get_or_create(
+        address=email_address,
+    )
+    return associate_email(
         profile=profile,
-        email=email,
+        email=email_object,
         type=type,
         preferred=preferred,
     )
```
[the remark this answers is not in the document]
Comment (marcelkornblum) [reply profiles/services/staff_sso.py:108]: I'd like to discuss this alongside the other bit I haven't resolved. I don't think I agree with this but would like to test my thinking

File: profiles/services/staff_sso.py [hunk 9]
```diff
@@ -1,106 +1,157 @@
...
-        )
-        update_staff_sso_email(
-            profile=staff_sso_profile,
-            email=email_object,
-            type=email["type"],
-            preferred=email["preferred"],
-        )
+    staff_sso_profile_email = StaffSSOProfileEmail.objects.get(
+        email=email, profile=profile
+    )
```
[the remark this answers is not in the document]
Comment (marcelkornblum) [reply profiles/services/staff_sso.py:140]: wrong type - should have been `Email`

**Review by CamLamb — APPROVED** [review]

**Pull request merged**